### ISSUE TITLE
Large discrepancies when creating Tensorrt engine in FP16 on transformer based model

### ISSUE THREAD

**Issue by zmy1116** [NVIDIA/TensorRT#1262]
## Description
Hello,

I'm working on creating trt engine for TimeSformer.
https://github.com/facebookresearch/TimeSformer

I build the network use TensorRT python API and I can confirm the engine produce reasonable result in FP32. However, the FP16 engine produces results that is far with expected output.

The main problem is that within each self attention block, the `gelu` function amplifies the discrepancies accumulated before, and since there are 12 blocks chained, eventually the discrepancy explode.

The table below shows normalized discrepancy (normalized err is defined as `norm(diff)/ norm(fp32_value)`) between FP16 and FP32 on 6 different steps for each of 12 self-attention blocks:
- after time self-attention
- after spatial self-attention
- after MLP - FC1
- after MLP - GELU
- after MLP - FC2
- output 

```
    T-attention  S-attention       FC1      GELU       FC2    output
0      0.000429     0.001589  0.001332  0.003152  0.003271  0.002415
1      0.002342     0.002767  0.001660  0.004542  0.004563  0.003427
2      0.003424     0.003747  0.002070  0.005539  0.006106  0.004365
3      0.004315     0.004391  0.002267  0.005573  0.007846  0.005714
4      0.005700     0.006732  0.003855  0.010994  0.023288  0.014126
5      0.013844     0.013440  0.007358  0.029005  0.058700  0.049154
6      0.049353     0.054302  0.102258  0.473737  0.535258  0.336202
7      0.335790     0.335395  0.139182  0.372451  0.702222  0.348630
8      0.348025     0.349228  0.183254  0.399796  0.538561  0.356199
9      0.355522     0.361621  0.401468  0.669973  0.805795  0.383704
10     0.381401     0.397778  0.706607  0.966223  0.910731  0.400721
11     0.441942     0.557325  0.280049  1.161975  1.132105  0.569143
```

You can see that GELU amplifies discrepancy from previous step 3x+ for every block, eventually the error goes to 100%+

## Environment
I used the ngc docker `docker pull nvcr.io/nvidia/tensorrt:21.05-py3`

**TensorRT Version**: 7.2.3.4 
**NVIDIA GPU**: Tesla T4  ,    AWS  g4dn machine 
**NVIDIA Driver Version**:  440.64.00
**CUDA Version**:  11.3
**CUDNN Version**:  
**Operating System**:  ubuntu
**Python Version (if applicable)**: 
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**: 

## Relevant Files
I have created the following repository to reproduce the issue 
https://github.com/zmy1116/timesfromer_trtfp16_error

## Steps To Reproduce

```
# clone repo
git clone https://github.com/zmy1116/timesfromer_trtfp16_error

cd timesfromer_trtfp16_error

# download pretrained weights from TimeSformer official repo
wget https://www.dropbox.com/s/g5t24we9gl5yk88/TimeSformer_divST_8x32_224_K400.pyth

# build 2 models in fp32 and fp16, get the outputs and do comparison
python conversions.py TimeSformer_divST_8x32_224_K400.pyth test_inputs.p result_fp32.p

python conversions.py TimeSformer_divST_8x32_224_K400.pyth test_inputs.p result_fp16.p --fp16_mode

# do comparison
python compare.py
```

you should see the same error table i presented at the beginning.

I think this is not a problem for specific model,   if raw scale of discrepancy is large, then gelu is going to amplfy it, after few blocks the results are certainly off.  I'm sort of surprised that you have not dealt this with BERT because I assume the same issue should happen on other transformer based model.

Thanks

**Comment by ttyio** [comment]
Hello @zmy1116 , we have a fix to remove the race condition in the FC + gelu fused kernel in 8.0, Could you take a try on 8.0EA to see if the large discrepancies still exist? thanks!

https://developer.nvidia.com/nvidia-tensorrt-download

**Comment by zmy1116** [comment]
@ttyio 

Using TRT 7.2,  I get the same discrepancies with gelu plugin and gelu written in native tensorrt api 

Using TRT 8.0,  the engine produces NAN on both FP16 and FP32,  I would have to dig further to see which line things explode but it looks like it has something to do with the qkvbert plugin 

Of course TRT 8.0 has some changes like fp16 and max_workspace_size are set on config not on builder.  So it may also me who made the mistake. 
I changed relevant part as following 
```
config=  builder.create_builder_config() 
config.max_workspace_size = 2 << 31 
if fp16_mode:
    config.set_flag(trt.BuilderFlag.FP16)
# builder.max_workspace_size = 2 << 31
#builder.fp16_mode = fp16_mode

...
 
engine = builder.build_engine(network, config)
```

**Comment by ttyio** [comment]
@zmy1116 
Have you tried 7.2 without marking `FC` before `gelu` as network output? this will prevent TRT from fuse them together. by enable verbose log, we can check if they are fused together, thanks!

**Comment by zmy1116** [comment]
@ttyio 

Without marking `FC1` I can fuse the FC1 with gelu but the discrepancy still persists
```
[TensorRT] VERBOSE: --------------- Timing Runner: mlp_fc1_6 + PWN(PWN(PWN(PWN(PWN(PWN((Unnamed Layer* 486) [Constant], (Unnamed Layer* 491) [ElementWise]), (Unnamed Layer* 487) [Constant] + (Unnamed Layer* 492) [ElementWise]), (Unnamed Layer* 493) [ElementWise]), (Unnamed Layer* 488) [Constant] + (Unnamed Layer* 494) [ElementWise] + (Unnamed Layer* 495) [Activation]), (Unnamed Layer* 489) [Constant] + (Unnamed Layer* 496) [ElementWise] + (Unnamed Layer* 490) [Constant] + (Unnamed Layer* 497) [ElementWise]), (Unnamed Layer* 498) [ElementWise]) (FusedConvActConvolution)

    T-attention  S-attention      GELU       FC2    output
0      0.000429     0.001589  0.002983  0.003209  0.002380
1      0.002307     0.002728  0.004360  0.004453  0.003365
2      0.003364     0.003691  0.005392  0.005981  0.004294
3      0.004246     0.004297  0.005406  0.007557  0.005502
4      0.005491     0.006469  0.010402  0.022191  0.013446
5      0.013185     0.012801  0.026469  0.053497  0.044972
6      0.045196     0.050229  0.545994  0.646239  0.405020
7      0.404189     0.403355  0.404985  1.068533  0.441110
8      0.439664     0.439896  0.446467  0.613117  0.457666
9      0.455490     0.462694  0.970165  1.380892  0.510623
10     0.503479     0.520674  0.863133  1.051000  0.520255
11     0.546814     0.638726  0.803358  0.647128  0.654663

```

**Comment by zmy1116** [comment]
@ttyio not sure if you are still looking on this issue but I can give you some updates on my end:

1. While gelu is a source that magnifies the error, **I do not think the source problem is at discrepancy at  FC + GELU** 

The way to prove this is to take the FP16 and FP32 values **before**  FC+GELU as inputs and simply run FC+GELU in pytorch with these 2 inputs

Below list one of the experiment I did,   from TensorRT FP16 and FP32 engine produce discripency around 0.044, matches exactly pytorch result.

If we believe pytorch FC + GELU produces exact value, then the discrepancy of FC+GELU is just magnifying the discrepancy of inputs, it has nothing to do with how FC and GELU are optimized /fused in TensorRT.

![fcgelu](https://user-images.githubusercontent.com/45111899/121463831-c3a06e00-c980-11eb-942e-6a37183f22ac.PNG)

2. Given 1, one experiment to do is to **not put gelu**, so instead of FC1-GELU-FC2 , I just have FC1 - FC2.     And indeed we still see large discrepancies at the end. 
I'm outputing the following entries per block:
- layer norm value before going in T attention
- T attention result
- S attention result
- MLP-FC1 
- MLP-FC2
- Output: patch features
- Output: class token
![exp1](https://user-images.githubusercontent.com/45111899/121464278-9b653f00-c981-11eb-8ea0-6fe5c0a4e143.PNG)

Interestingly, you see there is still large discrepancies at the end and the the discrepancy explodes starting at **7th block first layer normalization**

3. The next thing to look at is of course the layer normalization , I know there is a plugin for layer norm i have no idea how to use it alone so I wrote it myself.
```
def layer_norm(network, inputs, weights, bias, axes=4):
    # compute mean
    mean = network.add_reduce(inputs, trt.ReduceOperation.AVG, axes=axes, keep_dims=True).get_output(0)

    # compute diff
    diff = network.add_elementwise(inputs, mean, trt.ElementWiseOperation.SUB).get_output(0)
    
    # compute std
    POW = network.add_constant((1, ) * len(diff.shape), trt.Weights(np.ascontiguousarray([2.0], dtype=np.float32)))
    x = network.add_elementwise(diff, POW.get_output(0), trt.ElementWiseOperation.POW).get_output(0)
    x = network.add_reduce(x, trt.ReduceOperation.AVG, axes=axes, keep_dims=True).get_output(0)
    eps =  np.array([1e-6])
    eps = eps.astype(np.float32)
    x = network.add_scale(x, mode=trt.ScaleMode.ELEMENTWISE, shift=trt.Weights(eps)).get_output(0)
    std = network.add_unary(x, trt.UnaryOperation.SQRT).get_output(0)

    # compute normalized inputs
    normalized_inputs = network.add_elementwise(diff, std, trt.ElementWiseOperation.DIV).get_output(0)

    # weights bias 
    normalized_inputs = network.add_scale(normalized_inputs, mode=trt.ScaleMode.CHANNEL, scale=trt.Weights(weights),
                                    shift=trt.Weights(bias)).get_output(0)

    
    return mean, diff, std, normalized_inputs
```
You can see that here I returned 4 steps outputs, after some trial and error, I found that if I add the **diff** as engine output, the normalized error minimizes to 0.003
![exp2](https://user-images.githubusercontent.com/45111899/121464878-b71d1500-c982-11eb-8734-b73c5f521e19.PNG)


4. Now I add GELU back,   per block I expose the following entries as engine output:
- the layer norm diff before T transform
- the layer norm diff before S transform
- the layer norm diff before MLP 
- real output: patches features
- real output: classification token 
![exp8](https://user-images.githubusercontent.com/45111899/121465241-71ad1780-c983-11eb-874d-02f69ff5ea94.PNG)

You can see for the 2 outputs of interest, the error dropped below 0.05,    And the actual output I need is the last one for classification, the error is only 0.003, totally acceptable. 


With this setup,  I build the entire TimeSformer in FP32 and FP16, I output the folloiwng:
- 12 blocks per block the 3 diff from layer norm
- final classification output

I confirm the final result in fp32 and fp16 have discrepancy < 0.01 



To summarize, I think the core problem is within the layer norm I implemented, certain parts around the diff line got fused together and create non trivial discrepancy, and that discrepancy get magnifies through nonlinearity. 


Now I have no idea why the layer norm issue start to happen at the 7th block. 

What I present here fairly different than my repo.  I will update the repo later.


Also like I have mentioned before, trt 8.0 produces NAN, I have not had a chance to find out exactly which line cause the problem, not sure if you have the same thing.  

Thanks

**Comment by ttyio** [comment]
Cool @zmy1116 thanks for this analysis and sorry we are busy on 8.0 release these days, no bandwidth to check this yet.

Some thoughts from my side:
1. check the `sqrt` in layernorm, not sure if `FP16` provide enough mantissa bits here.
2. replace the input `dummy_video` with real data? low precision may only work on data with the same distribution as training data
3. do you know which layer generated the NAN?

Thanks!

**Comment by cszer** [comment]
@zmy1116 Hy man, please can you tell me about using QKV plugin after Linear in VIT Attention block. As i understand after permuting input i get [3,1,12,577,192] (self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4))
if input size is 384x384 , but in plugin docs authors say that shape is [S, B, 3 * E, 1, 1] . Does plugin work with this input dimensions ? With what values i need to feel  hiddenSize, numHeads, S when i initialize the plugin?

**Comment by cszer** [comment]
@zmy1116 As i see in your repo you used basic tensorrt plugin, seems good that it works

**Comment by zmy1116** [comment]
@ttyio  I have updated the repository to reflect recent discoveries.
https://github.com/zmy1116/timesfromer_trtfp16_error

To recap/summarize:
- I attempt to build TRT engine for Timesformer, I'm able to produce correct engine in FP32, but the FP16 engine produces incorrect result
- Initially I thought the issue is due to non-linearity (GELU) augment errors, and you have given multiple suggestions on that front. After more analysis, it appears that the problems is not at non-linearity, since changing to other non-linearity function or not having non-linearity function still produces incorrect result
- I think I narrow down the issue to the layer normalization block.   By exposing the `diff` from layer norm function, described below, the FP16 engine can produces correct results.
```
def layer_norm(network, inputs, weights, bias, block_num, ln_num, axes=4):
    # compute mean
    mean = network.add_reduce(inputs, trt.ReduceOperation.AVG, axes=axes, keep_dims=True)
    mean.name = '{0}_{1}_mean_inputs'.format(block_num, ln_num)
    mean = mean.get_output(0)

    # compute diff
    diff = network.add_elementwise(inputs, mean, trt.ElementWiseOperation.SUB)
    diff.name = '{0}_{1}_input_sub_mean'.format(block_num, ln_num)
    diff = diff.get_output(0)

    # compute std
    POW = network.add_constant((1,) * len(diff.shape), trt.Weights(np.ascontiguousarray([2.0], dtype=np.float32)))
    POW.name = '{0}_{1}_power_const'.format(block_num, ln_num)

    x = network.add_elementwise(diff, POW.get_output(0), trt.ElementWiseOperation.POW)
    x.name = '{0}_{1}_squared_diff'.format(block_num, ln_num)
    x = x.get_output(0)

    x = network.add_reduce(x, trt.ReduceOperation.AVG, axes=axes, keep_dims=True)
    x.name = '{0}_{1}_mean_squared_diff'.format(block_num, ln_num)
    x = x.get_output(0)
    eps = np.array([1e-6])
    eps = eps.astype(np.float32)
    x = network.add_scale(x, mode=trt.ScaleMode.ELEMENTWISE, shift=trt.Weights(eps))
    x.name = '{0}_{1}_add_eps'.format(block_num, ln_num)
    x = x.get_output(0)

    std = network.add_unary(x, trt.UnaryOperation.SQRT)
    std.name = '{0}_{1}_std'.format(block_num, ln_num)
    std = std.get_output(0)

    # compute normalized inputs
    normalized_inputs = network.add_elementwise(diff, std, trt.ElementWiseOperation.DIV)
    normalized_inputs.name = '{0}_{1}_diff_div_std'.format(block_num, ln_num)
    normalized_inputs = normalized_inputs.get_output(0)

    # weights bias
    normalized_inputs = network.add_scale(normalized_inputs, mode=trt.ScaleMode.CHANNEL, scale=trt.Weights(weights),
                                          shift=trt.Weights(bias))
    normalized_inputs.name = '{0}_{1}_rescale'.format(block_num, ln_num)
    normalized_inputs = normalized_inputs.get_output(0)
    return mean, diff, std, normalized_inputs
```

Since I don't see a layernorm layer in TensorRT, I wrote the layernorm layer above. 

There are total 12 self-attention blocks, each block consists of 3 layernorm functions, so FP16 engine exposing the 36 diffs produces correct results

I updated the repository to convert the full timesformer model, taking input of a raw `8x3x224x224` video inputs and outputs
 final probability and 12 blocks patch features by default,  if exposing LN-dff then 36 diffs are also parts of the outputs. 
 
 Without exposing LN-diff in FP16, we expect to see the following error values
 ```
Final probabilities difference: 1.0182867
Patch embedding diff after 0 S-T blocks: 0.004299811087548733
Patch embedding diff after 1 S-T blocks: 0.00536126596853137
Patch embedding diff after 2 S-T blocks: 0.00613724160939455
Patch embedding diff after 3 S-T blocks: 0.00694211944937706
Patch embedding diff after 4 S-T blocks: 0.009320137090981007
Patch embedding diff after 5 S-T blocks: 0.013058185577392578
Patch embedding diff after 6 S-T blocks: 0.4091333746910095
Patch embedding diff after 7 S-T blocks: 0.430844247341156
Patch embedding diff after 8 S-T blocks: 0.44067439436912537
Patch embedding diff after 9 S-T blocks: 0.46074366569519043
Patch embedding diff after 10 S-T blocks: 0.45822423696517944
Patch embedding diff after 11 S-T blocks: 0.5575547814369202
 ```
 
I totally understand you probably want me to narrow down the issue to smaller section. Without good knowledge of how TensorRT engine works or ability to directly tweak the codes behind the scene, this is the finest level I can narrow the issue at at this point. I think there is a more generalized problem here that probably apply to other transformer based models also. 

I think it would be beneficial if we can have an isolated layernorm layer/plugin, I know there is something similar in the plugin but I think that plugin does layernorm plus an addition..


Thanks

**Comment by ttyio** [comment]
Good job and thanks @zmy1116 for this excellent summary .

So FP16 only works when you manually build layer norm layers and use `--expose_ln_entries`, is this right?

When some layer is marked as output, by default the last layer force to use FP32 output, and the FP32 precision is propagated back until we hit first none weighted sum layer. I have try to check the build log, I see the layer norm run in different precision when  `--expose_ln_entries` used.

with `--expose_ln_entries`,  below layers are run under FP32 precision :

      [TensorRT] VERBOSE: Layer(Reformat): 1_2_input_sub_mean output reformatter 0, Tactic: 0, 1_2_input_sub_mean output to be reformatted 0[Half(768,1,1)] -> (Unnamed Layer* 105) [ElementWise]_output[Float(768,1,1)]
      [TensorRT] VERBOSE: Layer(PointWiseV2): PWN(1_2_power_const, 1_2_squared_diff), Tactic: 9, (Unnamed Layer* 105) [ElementWise]_output[Float(768,1,1)] -> (Unnamed Layer* 107) [ElementWise]_output[Float(768,1,1)]
      [TensorRT] VERBOSE: Layer(Reduce): 1_2_mean_squared_diff, Tactic: 3, (Unnamed Layer* 107) [ElementWise]_output[Float(768,1,1)] -> (Unnamed Layer* 108) [Reduce]_output[Float(1,1,1)]
      [TensorRT] VERBOSE: Layer(Scale): 1_2_add_eps, Tactic: 0, (Unnamed Layer* 108) [Reduce]_output[Float(1,1,1)] -> (Unnamed Layer* 109) [Scale]_output[Float(1,1,1)]
      [TensorRT] VERBOSE: Layer(Reformat): PWN(1_2_std, 1_2_diff_div_std) input reformatter 0, Tactic: 0, (Unnamed Layer* 109) [Scale]_output[Float(1,1,1)] -> PWN(1_2_std, 1_2_diff_div_std) reformatted input 0[Half(1,1,1)]

without  `--expose_ln_entries`, below layers are run under FP16 precision :

      [TensorRT] VERBOSE: Layer(ElementWise): 1_2_input_sub_mean, Tactic: 1, (Unnamed Layer* 103) [ElementWise]_output[Half(768,1,1)], (Unnamed Layer* 104) [Reduce]_output[Half(1,1,1)] -> (Unnamed Layer* 105) [ElementWise]_output[Half(768,1,1)]
      [TensorRT] VERBOSE: Layer(PointWiseV2): PWN(1_2_power_const, 1_2_squared_diff), Tactic: 9, (Unnamed Layer* 105) [ElementWise]_output[Half(768,1,1)] -> (Unnamed Layer* 107) [ElementWise]_output[Half(768,1,1)]
      [TensorRT] VERBOSE: Layer(Reduce): 1_2_mean_squared_diff, Tactic: 3, (Unnamed Layer* 107) [ElementWise]_output[Half(768,1,1)] -> (Unnamed Layer* 108) [Reduce]_output[Half(1,1,1)]
      [TensorRT] VERBOSE: Layer(Scale): 1_2_add_eps, Tactic: 0, (Unnamed Layer* 108) [Reduce]_output[Half(1,1,1)] -> (Unnamed Layer* 109) [Scale]_output[Half(1,1,1)]

Given FP16 only have 10 mantissa bits, it might not suit for some input data distribution because there are some `pow` in the layernorm. If so, it is not likely trt bug, but arithmetic issue that we have to run layernorm in FP32 precision.  What do you think?

Thanks.

**Comment by nvpohanh** [comment]
@zmy1116 Could you try the latest TRT 8.4 GA release and see if the issue still exists? If it does, we will debug it. Thanks

**Comment by ttyio** [comment]
Close since no activity for more than 3 weeks, thanks!
